### DISCUSSION TITLE
Detecting loss of focus on LineEdit

### DISCUSSION THREAD

**Discussion by maxenko** [slint-ui/slint#5110] [Q&A] [answered]
I've build my own widget wrapping LineEdit, which accepts text, int or float and responds to mouse-wheel for increments. Its what I need and it works well.

There is only one problem, and its with partial float input. 

User doesn't always press enter when they are done. They sometimes click off the LineEdit box into another widget (or tab over to next widget), which leaves input as valid parse value for a float, but incomplete. So the value will be left like 2. or 3.00, where it should be parsed into 2 or 3.

What can I use in slint to detect focus loss on a LineEdit so that I can do one final parse of a value the same way as I do now on accepted callback (when enter is pressed)?

**Comment by ogoffart** [comment, marked as answer]
This is not really possible. Although there exist hack as shown in issue #112
[Example](https://slintpad.com/?snippet=import+%7B+AboutSlint%2C+Button%2C+VerticalBox%2C+LineEdit%2C+SpinBox+%7D+from+%22std-widgets.slint%22%3B%0Aexport+component+Demo+%7B%0A++++VerticalBox+%7B%0A++++++++alignment%3A+start%3B%0A++++++++Text+%7B%0A++++++++++++text%3A+%22Hello+World%21%22%3B%0A++++++++++++font-size%3A+24px%3B%0A++++++++++++horizontal-alignment%3A+center%3B%0A++++++++%7D%0A++++++++HorizontalLayout+%7B+alignment%3A+center%3B+Button+%7B+text%3A+%22OK%21%22%3B+%7D+%7D%0A++++++++le+%3A%3D+LineEdit+%7B+edited+%3D%3E+%7B+was-edited+%3D+true%3B+%7D+%7D%0A++++++++SpinBox+%7B%7D%0A++++%7D%0A++++%2F%2F+workaround+for+https%3A%2F%2Fgithub.com%2Fslint-ui%2Fslint%2Fissues%2F112%0A++++property+%3Cbool%3E+was-edited%3B%0A++++if+%21le.has-focus+%26%26+was-edited%3A+Rectangle+%7B%0A++++++++init+%3D%3E+%7B+le.text+%3D+%22Lost+Focus%22%3B+was-edited+%3D+false%3B+%7D%0A++++%7D%0A%7D%0A&style=native)

**Reply by maxenko** [reply]
That's more or less a solution, I forgot its all immediate and re-evaluates on each frame.

**Reply by bjorn** [reply]
It's not immediate and doesn't re-evaluate each frame. That example works because the change in the `has-focus` property triggers a re-evaluation of the conditional expression. When it became true, a `Rectangle` is instantiated, triggering its `init` callback. This callback immediately triggers the removal of that rectangle by setting `was-edited` to false.

So each time `has-focus` becomes false and `was-edited` is true, there is exactly one chance to perform custom logic.

The preferred solution would likely be a `has-focus-changed` callback, but this hack works for now.

**Comment by maxenko** [comment]
> That example works because the change in the has-focus property triggers a re-evaluation of the conditional expression. 

Got it, thank you. Would have kept going with this wrong assumption if you didn't elaborate. 

https://snapshots.slint.dev/master/docs/slint/src/language/syntax/conditions
